---
hide:
  - navigation
---


**BioNeMo Framework** is a free-to-use collection of Python modules that offer access to optimized, pre-trained biomolecular models and enable workflows training and fine-tuning your own custom models. Capabilities span various workloads and therapeutic modalities, such as molecular generation and representation learning, protein structure prediction and representation learning, protein-ligand and protein-protein docking, and DNA/RNA/single-cell embedding.


<div class="grid cards" markdown>

-   :material-book-open-variant:{ .lg } __User Guide__

    ---

    Install BioNeMo and set up your environment to start accelerating your bioinformatics workflows.

    [Get Started](user-guide/getting-started){ .md-button .md-button }

-   :material-code-greater-than:{ .lg .middle } __API Reference__

    ---

    Access comprehensive documentation on BioNeMo's sub-packages, functions, and classes.

    [API Reference](API_reference/bionemo/core/api/){ .md-button .md-button }

-   :material-robot-outline:{ .lg } __Models__

    ---

    Explore detailed instructions and best practices for using BioNeMo in your research.

    [Explore Models](models){ .md-button .md-button }



-   :material-database-outline:{ .lg .middle } __Datasets__

    ---

    Join the BioNeMo community and learn how to contribute to this open-source project.

    [Explore Datasets](datasets){ .md-button .md-button }


</div>
