{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9a88d40b-64de-4661-b3b0-7a4b87ba0162",
   "metadata": {
    "tags": []
   },
   "source": [
    "# BioNeMo - Geneformer inferencing for single cell downstream tasks\n",
    "\n",
    "This tutorial showcases how to run the BioNeMo container, pre-train a geneformer model, and use it for inferencing downstream single cell tasks. At the end of this tutorial, a user will learn:\n",
    "- launching the BioNeMo container\n",
    "- Download data from czi to use for inference.\n",
    "- Convert AnnData files into the sparse CSR memmap format used by BioNeMo\n",
    "- Download a pretrained checkpoint\n",
    "- Restore the geneformer checkpoint and perform inference with the czi dataset.\n",
    "\n",
    "\n",
    "### Prerequisites:\n",
    "- BioNeMo Framework container is running (refer to the [Quickstart Guide](../quickstart-fw.md))\n",
    "- Familiarity with some components of the BioNeMo framework such as the [Models](../models/megamolbart.md) and [Inferencing](../inference-grpc-fw.md)\n",
    "\n",
    "\n",
    "#### Running the BioNeMo container\n",
    "\n",
    "This example has been built by launching the container in a local machine with 2 x A6000 RTX GPUs. Refer to specific instructions for [remote and multi-node launch]\n",
    "\n",
    "Once the container is launched, navigate to http://0.0.0.0:8888, http://localhost:8888, or the IP address of the workstation/node. A JupyterLab instance should show up.\n",
    "\n",
    "#### Copy this code and input files into JupyterLab\n",
    "\n",
    "In the launched JupyterLab, run the codes in a Jupyter notebook as provided in the code cells below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e47327a-6549-4ff4-a32a-a32d07a1e706",
   "metadata": {},
   "source": [
    "## Getting example single cell data and setting it up for inference\n",
    "\n",
    "First, we must acquire single cell training data for inference. To do this we will install the cellxgene-census api and download a small dataset. We use the example provided by the czi api examples page to download a single h5ad file. Generally, our workflow expects a collection of h5ad files to be used for pre-training. In this case, we restrict to 100k cells from a single dataset  to keep training time and downloading time small."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "056ca29a-ffa1-4381-83b6-cad9a7b0c4d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com\n",
      "Requirement already satisfied: cellxgene-census in /workspace/bionemo/.local/lib/python3.10/site-packages (1.13.0)\n",
      "Requirement already satisfied: tiledbsoma~=1.9.1 in /workspace/bionemo/.local/lib/python3.10/site-packages (from cellxgene-census) (1.9.5)\n",
      "Requirement already satisfied: anndata in /usr/local/lib/python3.10/dist-packages (from cellxgene-census) (0.10.6)\n",
      "Requirement already satisfied: numpy>=1.21 in /usr/local/lib/python3.10/dist-packages (from cellxgene-census) (1.24.4)\n",
      "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from cellxgene-census) (2.31.0)\n",
      "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.10/dist-packages (from cellxgene-census) (4.7.1)\n",
      "Requirement already satisfied: s3fs>=2021.06.1 in /workspace/bionemo/.local/lib/python3.10/site-packages (from cellxgene-census) (2024.3.1)\n",
      "Requirement already satisfied: aiobotocore<3.0.0,>=2.5.4 in /workspace/bionemo/.local/lib/python3.10/site-packages (from s3fs>=2021.06.1->cellxgene-census) (2.12.3)\n",
      "Requirement already satisfied: fsspec==2024.3.1 in /workspace/bionemo/.local/lib/python3.10/site-packages (from s3fs>=2021.06.1->cellxgene-census) (2024.3.1)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in /usr/local/lib/python3.10/dist-packages (from s3fs>=2021.06.1->cellxgene-census) (3.9.0)\n",
      "Requirement already satisfied: attrs>=22.2 in /usr/local/lib/python3.10/dist-packages (from tiledbsoma~=1.9.1->cellxgene-census) (23.1.0)\n",
      "Requirement already satisfied: numba>=0.58.0 in /workspace/bionemo/.local/lib/python3.10/site-packages (from tiledbsoma~=1.9.1->cellxgene-census) (0.59.1)\n",
      "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from tiledbsoma~=1.9.1->cellxgene-census) (1.5.3)\n",
      "Requirement already satisfied: pyarrow-hotfix in /usr/local/lib/python3.10/dist-packages (from tiledbsoma~=1.9.1->cellxgene-census) (0.6)\n",
      "Requirement already satisfied: scanpy>=1.9.2 in /usr/local/lib/python3.10/dist-packages (from tiledbsoma~=1.9.1->cellxgene-census) (1.9.8)\n",
      "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from tiledbsoma~=1.9.1->cellxgene-census) (1.11.1)\n",
      "Requirement already satisfied: somacore==1.0.10 in /workspace/bionemo/.local/lib/python3.10/site-packages (from tiledbsoma~=1.9.1->cellxgene-census) (1.0.10)\n",
      "Requirement already satisfied: tiledb~=0.27.0 in /workspace/bionemo/.local/lib/python3.10/site-packages (from tiledbsoma~=1.9.1->cellxgene-census) (0.27.1)\n",
      "Requirement already satisfied: pyarrow>=9.0.0 in /usr/local/lib/python3.10/dist-packages (from tiledbsoma~=1.9.1->cellxgene-census) (14.0.1)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /usr/local/lib/python3.10/dist-packages (from anndata->cellxgene-census) (1.5.1)\n",
      "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anndata->cellxgene-census) (1.1.3)\n",
      "Requirement already satisfied: h5py>=3.1 in /usr/local/lib/python3.10/dist-packages (from anndata->cellxgene-census) (3.10.0)\n",
      "Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from anndata->cellxgene-census) (8.4.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from anndata->cellxgene-census) (23.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->cellxgene-census) (3.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->cellxgene-census) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->cellxgene-census) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->cellxgene-census) (2023.7.22)\n",
      "Requirement already satisfied: botocore<1.34.70,>=1.34.41 in /workspace/bionemo/.local/lib/python3.10/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census) (1.34.69)\n",
      "Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in /usr/local/lib/python3.10/dist-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census) (1.14.1)\n",
      "Requirement already satisfied: aioitertools<1.0.0,>=0.5.1 in /workspace/bionemo/.local/lib/python3.10/site-packages (from aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census) (0.11.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census) (6.0.4)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census) (1.9.2)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census) (1.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census) (1.3.1)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->s3fs>=2021.06.1->cellxgene-census) (4.0.3)\n",
      "Requirement already satisfied: llvmlite<0.43,>=0.42.0dev0 in /workspace/bionemo/.local/lib/python3.10/site-packages (from numba>=0.58.0->tiledbsoma~=1.9.1->cellxgene-census) (0.42.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /usr/local/lib/python3.10/dist-packages (from pandas->tiledbsoma~=1.9.1->cellxgene-census) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->tiledbsoma~=1.9.1->cellxgene-census) (2023.3)\n",
      "Requirement already satisfied: joblib in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (1.2.0)\n",
      "Requirement already satisfied: matplotlib>=3.6 in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (3.8.0)\n",
      "Requirement already satisfied: networkx>=2.3 in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (2.6.3)\n",
      "Requirement already satisfied: patsy in /workspace/bionemo/.local/lib/python3.10/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.5.6)\n",
      "Requirement already satisfied: scikit-learn>=0.24 in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (1.2.0)\n",
      "Requirement already satisfied: seaborn>=0.13.0 in /workspace/bionemo/.local/lib/python3.10/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.13.2)\n",
      "Requirement already satisfied: session-info in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (1.0.0)\n",
      "Requirement already satisfied: statsmodels>=0.10.0rc2 in /workspace/bionemo/.local/lib/python3.10/site-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.14.1)\n",
      "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (4.66.1)\n",
      "Requirement already satisfied: umap-learn>=0.3.10 in /usr/local/lib/python3.10/dist-packages (from scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.5.5)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /usr/local/lib/python3.10/dist-packages (from botocore<1.34.70,>=1.34.41->aiobotocore<3.0.0,>=2.5.4->s3fs>=2021.06.1->cellxgene-census) (1.0.1)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (1.1.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /workspace/bionemo/.local/lib/python3.10/site-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (1.4.5)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (10.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.6->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (3.1.1)\n",
      "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.1->pandas->tiledbsoma~=1.9.1->cellxgene-census) (1.16.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.24->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (3.2.0)\n",
      "Requirement already satisfied: pynndescent>=0.5 in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.3.10->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.5.11)\n",
      "Requirement already satisfied: stdlib-list in /usr/local/lib/python3.10/dist-packages (from session-info->scanpy>=1.9.2->tiledbsoma~=1.9.1->cellxgene-census) (0.10.0)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.0\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpython -m pip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install cellxgene-census"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "24f579ac-0207-4e07-8f34-dac0cd955c53",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Below are paths required for setting up pre-training and inference.\n",
    "tutorial_data_dir = \"/workspace/bionemo/data/singlecell_tutorial/download_anndata\"\n",
    "tutorial_processed_dir = \"/workspace/bionemo/data/singlecell_tutorial/processed_data\"\n",
    "tutorial_output_dir = \"/workspace/bionemo/data/singlecell_tutorial/inference_output\"\n",
    "tutorial_output_inference_pickle = f\"{tutorial_output_dir}/human_covid19_bcells_pretrained_inference.pkl\"\n",
    "demo_data_download_path = f\"{tutorial_data_dir}/human_covid19_bcells.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "88bd2844-0251-476e-9e3f-8d4e89474f33",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!mkdir -p {tutorial_data_dir}\n",
    "!mkdir -p {tutorial_processed_dir}\n",
    "!mkdir -p {tutorial_output_dir}\n",
    "!rm -f {tutorial_output_inference_pickle}  # clean this up if it's already there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6ccf03f0-1590-434e-a314-8ff0a02f5ef2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.10/dist-packages/anndata/_core/anndata.py:1209: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.10/dist-packages/anndata/_core/anndata.py:1209: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.10/dist-packages/anndata/_core/anndata.py:1209: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.10/dist-packages/anndata/_core/anndata.py:1209: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.10/dist-packages/anndata/_core/anndata.py:1209: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n",
      "/usr/local/lib/python3.10/dist-packages/anndata/_core/anndata.py:1209: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  df[key] = c\n"
     ]
    }
   ],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "with cellxgene_census.open_soma(census_version=\"2023-12-15\") as census:\n",
    "    filter1 = \"cell_type == 'B cell' and tissue_general == 'lung' and disease == 'COVID-19' and is_primary_data == True\"\n",
    "\n",
    "    adata = cellxgene_census.get_anndata(\n",
    "        census = census,\n",
    "        organism = \"Homo sapiens\",\n",
    "        obs_value_filter = filter1,\n",
    "        column_names = {\"obs\": [\"assay\", \"cell_type\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\"]},\n",
    "    )\n",
    "\n",
    "    adata[:100000].write(demo_data_download_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b060e73b-a3f9-4315-ae91-c23b9a29f490",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 jstjohn domain-users 27M Apr 29 16:01 /workspace/bionemo/data/singlecell_tutorial/download_anndata/human_covid19_bcells.h5ad\n"
     ]
    }
   ],
   "source": [
    "!ls -laht {demo_data_download_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1812f0e-7179-4140-aa9c-8109813f3e5e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1 files\n",
      "Starting to create memmap files...\n",
      "Metadata already exists, loading...\n",
      "Writing data into memmaps to /workspace/bionemo/data/singlecell_tutorial/processed_data...\n",
      "Merging AnnData into numpy memaps...: 100%|███████| 1/1 [00:00<00:00,  7.35it/s]\n",
      "Saving dataframe ...\n",
      "Done creating dataset ...\n"
     ]
    }
   ],
   "source": [
    "!python /workspace/bionemo/bionemo/data/singlecell/sc_memmap.py \\\n",
    "  --data-path {tutorial_data_dir} \\\n",
    "  --save-path {tutorial_processed_dir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bd212acb-9695-4ec4-bff6-8669693118f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 18M\n",
      "-rw-r--r-- 1 jstjohn domain-users 107K Apr 29 16:01 features.csv\n",
      "-rw-r--r-- 1 jstjohn domain-users 8.5M Apr 29 16:01 gene_expression_ind.npy\n",
      "-rw-r--r-- 1 jstjohn domain-users  19K Apr 29 16:01 gene_expression_ptr.npy\n",
      "-rw-r--r-- 1 jstjohn domain-users 8.5M Apr 29 16:01 gene_expression_data.npy\n",
      "drwxr-xr-x 5 jstjohn domain-users 4.0K Apr 26 18:55 ..\n",
      "drwxr-xr-x 2 jstjohn domain-users 4.0K Apr 26 18:46 .\n",
      "-rw-r--r-- 1 jstjohn domain-users 764K Apr 26 18:46 metadata.json\n"
     ]
    }
   ],
   "source": [
    "!ls -laht {tutorial_processed_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d23771e9-0dd6-414e-8cd0-5cd9a9dd4a6b",
   "metadata": {},
   "source": [
    "# Running inference.\n",
    "\n",
    "Now we will use a pretrained nemo file and start there for inference. For this tutorial we'll use the large variant of our pretrained checkpoint, the 103M parameter variant based on BERT-base with enhancements from the geneformer publications."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f474c69-743d-41b0-ab08-7420c0edf45b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: aws s3 cp s3://bionemo-ci/models/geneformer-31M-240430-step-73505-wandb-iyxsxz2m.nemo models/ --endpoint-url https://pbss.s8k.io\n",
      "\n",
      "\n",
      "Done.\n",
      "Running command: mkdir -p models/singlecell/geneformer && ln -sf ../../geneformer-31M-240430-step-73505-wandb-iyxsxz2m.nemo models/singlecell/geneformer/geneformer-31M-240430-large.nemo\n",
      "\n",
      "\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "!cd /workspace/bionemo/ && python download_models.py  --download_dir models geneformer_31M_240430_large #--source pbss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b11a330-b8f6-491a-b608-fe47e78f83ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pretrained_nemo_file = '/workspace/bionemo/models/singlecell/geneformer/geneformer-31M-240430-large.nemo'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "269a327d-a5b1-49c9-96b3-141ae99baeb3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NeMo W 2024-04-29 16:06:02 nemo_logging:349] /usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.\n",
      "      self.pid = os.fork()\n",
      "    \n",
      "[NeMo I 2024-04-29 16:06:03 megatron_hiddens:110] Registered hidden transform sampled_var_cond_gaussian at bionemo.model.core.hiddens_support.SampledVarGaussianHiddenTransform\n",
      "[NeMo I 2024-04-29 16:06:03 megatron_hiddens:110] Registered hidden transform interp_var_cond_gaussian at bionemo.model.core.hiddens_support.InterpVarGaussianHiddenTransform\n",
      "[NeMo W 2024-04-29 16:06:03 nemo_logging:349] /usr/local/lib/python3.10/dist-packages/hydra/_internal/hydra.py:119: UserWarning: Future Hydra versions will no longer change working directory at job runtime by default.\n",
      "    See https://hydra.cc/docs/1.2/upgrades/1.1_to_1.2/changes_to_job_working_dir/ for more information.\n",
      "      ret = run_job(\n",
      "    \n",
      "[NeMo I 2024-04-29 16:06:03 loading:31] \n",
      "    \n",
      "    ************** Experiment configuration ***********\n",
      "[NeMo I 2024-04-29 16:06:03 loading:32] \n",
      "    name: geneformer_inference\n",
      "    desc: Minimum configuration for initializing a Geneformer model for inference.\n",
      "    trainer:\n",
      "      precision: 16-mixed\n",
      "      devices: 1\n",
      "      num_nodes: 1\n",
      "      accelerator: gpu\n",
      "      logger: false\n",
      "    exp_manager:\n",
      "      explicit_log_dir: null\n",
      "      exp_dir: /workspace/bionemo/data/singlecell_tutorial/inference_output\n",
      "      name: ${name}\n",
      "      create_checkpoint_callback: false\n",
      "    model:\n",
      "      micro_batch_size: ${model.data.batch_size}\n",
      "      downstream_task:\n",
      "        restore_from_path: /workspace/bionemo/models/singlecell/geneformer/geneformer-31M-240430-large.nemo\n",
      "        outputs:\n",
      "        - embeddings\n",
      "        - hiddens\n",
      "      data:\n",
      "        num_workers: 4\n",
      "        batch_size: 8\n",
      "        dataset_path: /workspace/bionemo/data/singlecell_tutorial/processed_data\n",
      "        output_fname: /workspace/bionemo/data/singlecell_tutorial/inference_output/human_covid19_bcells_pretrained_inference.pkl\n",
      "        index_mapping_dir: null\n",
      "        data_fields_map:\n",
      "          sequence: sequence\n",
      "          id: id\n",
      "        data_impl: geneformer\n",
      "        data_impl_kwargs:\n",
      "          csv_fields_mmap:\n",
      "            newline_int: 10\n",
      "            header_lines: 1\n",
      "            workers: null\n",
      "            sort_dataset_paths: false\n",
      "            data_sep: ','\n",
      "            data_fields:\n",
      "              id: 0\n",
      "              sequence: 1\n",
      "          fasta_fields_mmap:\n",
      "            data_fields:\n",
      "              id: 0\n",
      "              sequence: 1\n",
      "        dynamic_padding: true\n",
      "      post_process: false\n",
      "    target: bionemo.model.singlecell.geneformer.model.GeneformerModel\n",
      "    infer_target: bionemo.model.singlecell.geneformer.infer.GeneformerInference\n",
      "    formatters:\n",
      "      simple:\n",
      "        format: '[%(asctime)s][%(name)s][%(levelname)s] - %(message)s'\n",
      "    handlers:\n",
      "      console:\n",
      "        class: logging.StreamHandler\n",
      "        formatter: simple\n",
      "        stream: ext://sys.stdout\n",
      "      file:\n",
      "        class: logging.FileHandler\n",
      "        formatter: simple\n",
      "        filename: /logs/inference.log\n",
      "    root:\n",
      "      level: INFO\n",
      "      handlers:\n",
      "      - console\n",
      "    disable_existing_loggers: false\n",
      "    \n",
      "[NeMo I 2024-04-29 16:06:03 utils:333] Restoring model from /workspace/bionemo/models/singlecell/geneformer/geneformer-31M-240430-large.nemo\n",
      "[NeMo I 2024-04-29 16:06:03 utils:337] Loading model class: bionemo.model.singlecell.geneformer.model.GeneformerModel\n",
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "IPU available: False, using: 0 IPUs\n",
      "HPU available: False, using: 0 HPUs\n",
      "[NeMo I 2024-04-29 16:06:04 exp_manager:394] Experiments will be logged at /workspace/bionemo/data/singlecell_tutorial/inference_output/geneformer_inference/2024-04-29_16-06-04\n",
      "[NeMo I 2024-04-29 16:06:04 exp_manager:835] TensorboardLogger has been set up\n",
      "[NeMo I 2024-04-29 16:06:04 utils:306] \n",
      "    \n",
      "    ************** Trainer configuration ***********\n",
      "[NeMo I 2024-04-29 16:06:04 utils:307] \n",
      "    name: geneformer_inference\n",
      "    desc: Minimum configuration for initializing a Geneformer model for inference.\n",
      "    trainer:\n",
      "      precision: 16-mixed\n",
      "      devices: 1\n",
      "      num_nodes: 1\n",
      "      accelerator: gpu\n",
      "      logger: false\n",
      "      accumulate_grad_batches: 1\n",
      "    exp_manager:\n",
      "      explicit_log_dir: null\n",
      "      exp_dir: /workspace/bionemo/data/singlecell_tutorial/inference_output\n",
      "      name: ${name}\n",
      "      create_checkpoint_callback: false\n",
      "    model:\n",
      "      tokenizer:\n",
      "        vocab_file: nemo:d2fa1c61923d49fdb5b582ed49f58a38_geneformer.vocab\n",
      "      micro_batch_size: ${model.data.batch_size}\n",
      "      tensor_model_parallel_size: 1\n",
      "      pipeline_model_parallel_size: 1\n",
      "      use_flash_attention: true\n",
      "      seq_length: 2048\n",
      "      encoder_seq_length: 2048\n",
      "      max_position_embeddings: 2048\n",
      "      num_layers: 12\n",
      "      hidden_size: 768\n",
      "      ffn_hidden_size: 3072\n",
      "      num_attention_heads: 12\n",
      "      init_method_std: 0.02\n",
      "      hidden_dropout: 0.02\n",
      "      kv_channels: null\n",
      "      apply_query_key_layer_scaling: true\n",
      "      layernorm_epsilon: 1.0e-12\n",
      "      make_vocab_size_divisible_by: 128\n",
      "      pre_process: true\n",
      "      post_process: false\n",
      "      bert_binary_head: false\n",
      "      resume_from_checkpoint: null\n",
      "      masked_softmax_fusion: true\n",
      "      native_amp_init_scale: 4294967296\n",
      "      native_amp_growth_interval: 1000\n",
      "      fp32_residual_connection: true\n",
      "      fp16_lm_cross_entropy: false\n",
      "      seed: 1234\n",
      "      use_cpu_initialization: false\n",
      "      onnx_safe: false\n",
      "      activations_checkpoint_method: null\n",
      "      activations_checkpoint_num_layers: 1\n",
      "      data:\n",
      "        data_impl: geneformer\n",
      "        probabilistic_dirichlet_sampling_train: false\n",
      "        dataset_path: /workspace/bionemo/data/singlecell_tutorial/processed_data\n",
      "        data_prefix: ''\n",
      "        shuffle: true\n",
      "        dataset: /\n",
      "        medians_file: nemo:b6195b210ba34538a9c2b9c783e7bfa9_medians.json\n",
      "        index_mapping_dir: null\n",
      "        skip_warmup: true\n",
      "        index_mapping_type: memmap\n",
      "        train_ratio: 0.98\n",
      "        val_ratio: 0.01\n",
      "        num_workers: 4\n",
      "        dataloader_type: single\n",
      "        seq_length: 2048\n",
      "        seed: 1234\n",
      "        dynamic_padding: true\n",
      "        micro_batch_size: 16\n",
      "        batch_size: 8\n",
      "        output_fname: /workspace/bionemo/data/singlecell_tutorial/inference_output/human_covid19_bcells_pretrained_inference.pkl\n",
      "        data_fields_map:\n",
      "          sequence: sequence\n",
      "          id: id\n",
      "        data_impl_kwargs:\n",
      "          csv_fields_mmap:\n",
      "            newline_int: 10\n",
      "            header_lines: 1\n",
      "            workers: null\n",
      "            sort_dataset_paths: false\n",
      "            data_sep: ','\n",
      "            data_fields:\n",
      "              id: 0\n",
      "              sequence: 1\n",
      "          fasta_fields_mmap:\n",
      "            data_fields:\n",
      "              id: 0\n",
      "              sequence: 1\n",
      "      optim:\n",
      "        name: fused_adam\n",
      "        lr: 0.001\n",
      "        weight_decay: 0.1\n",
      "        betas:\n",
      "        - 0.9\n",
      "        - 0.999\n",
      "        sched:\n",
      "          name: CosineAnnealing\n",
      "          warmup_steps: 1000\n",
      "          constant_steps: 4000\n",
      "          max_steps: 100000\n",
      "          min_lr: 2.0e-05\n",
      "      precision: bf16-mixed\n",
      "      attention_dropout: 0.02\n",
      "      activation: relu\n",
      "      global_batch_size: 8\n",
      "      target: bionemo.model.singlecell.geneformer.model.GeneformerModel\n",
      "      nemo_version: 1.22.0\n",
      "      downstream_task:\n",
      "        restore_from_path: /workspace/bionemo/models/singlecell/geneformer/geneformer-31M-240430-large.nemo\n",
      "        outputs:\n",
      "        - embeddings\n",
      "        - hiddens\n",
      "    target: bionemo.model.singlecell.geneformer.model.GeneformerModel\n",
      "    infer_target: bionemo.model.singlecell.geneformer.infer.GeneformerInference\n",
      "    formatters:\n",
      "      simple:\n",
      "        format: '[%(asctime)s][%(name)s][%(levelname)s] - %(message)s'\n",
      "    handlers:\n",
      "      console:\n",
      "        class: logging.StreamHandler\n",
      "        formatter: simple\n",
      "        stream: ext://sys.stdout\n",
      "      file:\n",
      "        class: logging.FileHandler\n",
      "        formatter: simple\n",
      "        filename: /logs/inference.log\n",
      "    root:\n",
      "      level: INFO\n",
      "      handlers:\n",
      "      - console\n",
      "    disable_existing_loggers: false\n",
      "    \n",
      "[NeMo W 2024-04-29 16:06:05 modelPT:251] You tried to register an artifact under config key=tokenizer.vocab_file but an artifact for it has already been registered.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: context_parallel_size in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: virtual_pipeline_model_parallel_size in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: sequence_parallel in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: expert_model_parallel_size in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: gradient_accumulation_fusion in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_overlap in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_split_ag in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_split_rs in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_bulk_wgrad in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_bulk_dgrad in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: finalize_model_grads_func in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: overlap_p2p_comm in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: batch_p2p_comm in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: pipeline_model_parallel_split_rank in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: barrier_with_L1_time in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:234] Rank 0 has data parallel group: [0]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:237] All data parallel group ranks: [[0]]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:238] Ranks 0 has data parallel rank: 0\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:246] Rank 0 has model parallel group: [0]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:247] All model parallel group ranks: [[0]]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:257] Rank 0 has tensor model parallel group: [0]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:261] All tensor model parallel group ranks: [[0]]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:262] Rank 0 has tensor model parallel rank: 0\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:276] Rank 0 has pipeline model parallel group: [0]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:288] Rank 0 has embedding group: [0]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:294] All pipeline model parallel group ranks: [[0]]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:295] Rank 0 has pipeline model parallel rank 0\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:296] All embedding group ranks: [[0]]\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_init:297] Rank 0 has embedding rank: 0\n",
      "24-04-29 16:06:05 - PID:15854 - rank:(0, 0, 0, 0) - microbatches.py:39 - INFO - setting number of micro-batches to constant 1\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: context_parallel_size in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: virtual_pipeline_model_parallel_size in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: sequence_parallel in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: expert_model_parallel_size in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: gradient_accumulation_fusion in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_overlap in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_split_ag in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_split_rs in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_bulk_wgrad in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: tp_comm_bulk_dgrad in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: finalize_model_grads_func in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: overlap_p2p_comm in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: batch_p2p_comm in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: pipeline_model_parallel_split_rank in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 megatron_base_model:821] The model: GeneformerModel() does not have field.name: barrier_with_L1_time in its cfg. Add this key to cfg or config_mapping to make to make it configurable.\n",
      "[NeMo W 2024-04-29 16:06:05 modelPT:251] You tried to register an artifact under config key=tokenizer.vocab_file but an artifact for it has already been registered.\n",
      "[NeMo I 2024-04-29 16:06:05 megatron_base_model:315] Padded vocab_size: 21376, original vocab_size: 21250, dummy tokens: 126.\n",
      "[NeMo I 2024-04-29 16:06:05 nlp_overrides:752] Model GeneformerModel was successfully restored from /workspace/bionemo/models/singlecell/geneformer/geneformer-31M-240430-large.nemo.\n",
      "[NeMo I 2024-04-29 16:06:05 utils:474] DDP is not initialized. Initializing...\n",
      "Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1\n",
      "----------------------------------------------------------------------------------------------------\n",
      "distributed_backend=nccl\n",
      "All distributed processes registered. Starting with 1 processes\n",
      "----------------------------------------------------------------------------------------------------\n",
      "\n",
      "[NeMo W 2024-04-29 16:06:05 nemo_logging:349] /usr/local/lib/python3.10/dist-packages/apex/transformer/pipeline_parallel/utils.py:81: UserWarning: This function is only for unittest\n",
      "      warnings.warn(\"This function is only for unittest\")\n",
      "    \n",
      "[NeMo W 2024-04-29 16:06:06 nemo_logging:349] /usr/local/lib/python3.10/dist-packages/nemo/collections/nlp/modules/common/megatron/fused_bias_dropout_add.py:70: UserWarning: nvfuser integration in TorchScript is deprecated. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/interface.cpp:235.)\n",
      "      return bias_dropout_add_fused_inference_(*args)\n",
      "    \n",
      "[NeMo I 2024-04-29 16:06:06 loading:43] \n",
      "    \n",
      "    ************** Restored model configuration ***********\n",
      "[NeMo I 2024-04-29 16:06:06 loading:44] \n",
      "    tokenizer:\n",
      "      vocab_file: /tmp/tmpqz9a_o45/d2fa1c61923d49fdb5b582ed49f58a38_geneformer.vocab\n",
      "    micro_batch_size: 8\n",
      "    tensor_model_parallel_size: 1\n",
      "    pipeline_model_parallel_size: 1\n",
      "    use_flash_attention: true\n",
      "    seq_length: 2048\n",
      "    encoder_seq_length: 2048\n",
      "    max_position_embeddings: 2048\n",
      "    num_layers: 12\n",
      "    hidden_size: 768\n",
      "    ffn_hidden_size: 3072\n",
      "    num_attention_heads: 12\n",
      "    init_method_std: 0.02\n",
      "    hidden_dropout: 0.02\n",
      "    kv_channels: null\n",
      "    apply_query_key_layer_scaling: true\n",
      "    layernorm_epsilon: 1.0e-12\n",
      "    make_vocab_size_divisible_by: 128\n",
      "    pre_process: true\n",
      "    post_process: false\n",
      "    bert_binary_head: false\n",
      "    resume_from_checkpoint: null\n",
      "    masked_softmax_fusion: true\n",
      "    native_amp_init_scale: 4294967296\n",
      "    native_amp_growth_interval: 1000\n",
      "    fp32_residual_connection: true\n",
      "    fp16_lm_cross_entropy: false\n",
      "    seed: 1234\n",
      "    use_cpu_initialization: false\n",
      "    onnx_safe: false\n",
      "    activations_checkpoint_method: null\n",
      "    activations_checkpoint_num_layers: 1\n",
      "    data:\n",
      "      data_impl: geneformer\n",
      "      probabilistic_dirichlet_sampling_train: false\n",
      "      dataset_path: /workspace/bionemo/data/singlecell_tutorial/processed_data\n",
      "      data_prefix: ''\n",
      "      shuffle: true\n",
      "      dataset: /\n",
      "      medians_file: nemo:b6195b210ba34538a9c2b9c783e7bfa9_medians.json\n",
      "      index_mapping_dir: null\n",
      "      skip_warmup: true\n",
      "      index_mapping_type: memmap\n",
      "      train_ratio: 0.98\n",
      "      val_ratio: 0.01\n",
      "      num_workers: 4\n",
      "      dataloader_type: single\n",
      "      seq_length: 2048\n",
      "      seed: 1234\n",
      "      dynamic_padding: true\n",
      "      micro_batch_size: 16\n",
      "      batch_size: 8\n",
      "      output_fname: /workspace/bionemo/data/singlecell_tutorial/inference_output/human_covid19_bcells_pretrained_inference.pkl\n",
      "      data_fields_map:\n",
      "        sequence: sequence\n",
      "        id: id\n",
      "      data_impl_kwargs:\n",
      "        csv_fields_mmap:\n",
      "          newline_int: 10\n",
      "          header_lines: 1\n",
      "          workers: null\n",
      "          sort_dataset_paths: false\n",
      "          data_sep: ','\n",
      "          data_fields:\n",
      "            id: 0\n",
      "            sequence: 1\n",
      "        fasta_fields_mmap:\n",
      "          data_fields:\n",
      "            id: 0\n",
      "            sequence: 1\n",
      "    optim:\n",
      "      name: fused_adam\n",
      "      lr: 0.001\n",
      "      weight_decay: 0.1\n",
      "      betas:\n",
      "      - 0.9\n",
      "      - 0.999\n",
      "      sched:\n",
      "        name: CosineAnnealing\n",
      "        warmup_steps: 1000\n",
      "        constant_steps: 4000\n",
      "        max_steps: 100000\n",
      "        min_lr: 2.0e-05\n",
      "    precision: 16-mixed\n",
      "    attention_dropout: 0.02\n",
      "    activation: relu\n",
      "    global_batch_size: 8\n",
      "    target: bionemo.model.singlecell.geneformer.model.GeneformerModel\n",
      "    nemo_version: 1.22.0\n",
      "    downstream_task:\n",
      "      restore_from_path: /workspace/bionemo/models/singlecell/geneformer/geneformer-31M-240430-large.nemo\n",
      "      outputs:\n",
      "      - embeddings\n",
      "      - hiddens\n",
      "    \n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]\n",
      "[NeMo W 2024-04-29 16:06:07 nemo_logging:349] /usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.\n",
      "      self.pid = os.fork()\n",
      "    \n",
      "Predicting DataLoader 0: 100%|████████████████| 290/290 [00:55<00:00,  5.24it/s]\n",
      "[NeMo I 2024-04-29 16:07:03 run_inference:50] Collecting results from all GPUs...\n",
      "[NeMo I 2024-04-29 16:07:03 infer:73] Saving 2313 samples to /workspace/bionemo/data/singlecell_tutorial/inference_output/human_covid19_bcells_pretrained_inference.pkl\n"
     ]
    }
   ],
   "source": [
    "!python /workspace/bionemo/bionemo/model/infer.py \\\n",
    "  --config-dir /workspace/bionemo/examples/singlecell/geneformer/conf \\\n",
    "  --config-name infer \\\n",
    "  ++model.downstream_task.restore_from_path={pretrained_nemo_file} \\\n",
    "  ++model.data.batch_size=8 \\\n",
    "  ++model.data.dataset_path={tutorial_processed_dir} \\\n",
    "  ++exp_manager.exp_dir={tutorial_output_dir} \\\n",
    "  ++model.data.output_fname={tutorial_output_inference_pickle} "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1b9a425-850d-4027-9160-5e659c163604",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Load inference result and cluster with UMAP.\n",
    "Now we will inspect our result. First, we expect there to be one prediction for each cell, we can compare the shape of the anndata object to the predictions produced by our model. After this, we can simply pass our embeddings into umap, and view the result! In this case its a very poorly trained model with very few cells, so keep expectations low!\n",
    "\n",
    "The inference_results pickle file contains one set of hiddens and embeddings for each cell. The hiddens contain the embedding per-token, whereas the embeddings contain the mean embedding for all gene tokens with special tokens (CLS, MASK, etc) removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2895e979-1652-49f2-96e9-b63b66191ada",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2313,\n",
       " dict_keys(['text', 'types', 'padding_mask', 'labels', 'loss_mask', 'is_random', 'hiddens', 'embeddings']))"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle\n",
    "with open(tutorial_output_inference_pickle, 'rb') as inference_handle:\n",
    "    inference_results = pickle.load(inference_handle)\n",
    "len(inference_results), inference_results[0].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "621decfb-2fd1-44dd-86e5-69e7ddcb8235",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(768,)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inference_results[0]['embeddings'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d80061f4-3311-4246-9640-177d4ca3b9a7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import umap\n",
    "reducer = umap.UMAP()\n",
    "embedding = reducer.fit_transform([x['embeddings'] for x in inference_results])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "704d37ad-ac98-4868-9249-bd2ada17a8d7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2313, 2)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embedding.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5ea88b8a-bb3c-4058-93a0-9afb0dfedfa5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7fc196554af0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAGdCAYAAACyzRGfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABGHElEQVR4nO3db2xU55n//8+ZP4yn/oEHJwQyLpkQt9AEasPSDa6SjWiSBli+GNRK3xbVu2lVdVeWNqibRRVpQlwX0rBdVEVkI/Jo1a0soe4+WELkbclushZKNg7fsoDrJCWJl7qpHUgKjJ24YzN/zu/B+D4+M54xNthnxvb7JVm2Z87M3B6Mz3Wu+7qv27Jt2xYAAIBHfKUeAAAAmF8IPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcCpR5Avkwmo/7+fi1cuFCWZZV6OAAAYBJs29bHH3+saDQqn2/i3EbZBR/9/f1avnx5qYcBAACuw/vvv69Pf/rTEx5TdsHHwoULJWUHv2jRohKPBgAATMbg4KCWL1/unMcnUnbBh5lqWbRoEcEHAACzzGRKJig4BQAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4ACahrbNX9xx4RW2dvaUeCgDMepZt23apB+E2ODioqqoqDQwMsKstSq6ts1eHO3o0NJJSPJF0bjd7Nm6rj+rQznVq6+zVwePnNDSSUipjO7cDwHwxlfM3wQcwgXsOvKK+eGLCY8JBv4aTaeX/R4qEg9q9aZWaGmIzN0AAKBNTOX8z7QJMYH1s8TWPSRQIPCQpnkjq4PFz0z8oAJjlAqUeAFDOTvVeKXi7mXYJ+LJfJTOFE4gjqfRMDAsAZjWCD6CAXUdOq72rX6ujVZKk5o21kqQnjnZLkpPpuGVRxbh6ELdQwD/jYwWA2YbgAxhlikZHUmklkhlJ0pv9A+p5eqtzzFPtbzn3SdKHHw8rlR6f9QgHfaquDDlBCwBgDMEHMOrJF7qVP3tiMh9SNjhxBx6SlCwQeEjSl+9axmoXACiC4AMYVahs49d9A2p89lW92T8gn2WNPyCPpeyUTLFaEQAAq10AR02kwvl6tI5UtqSuvgGl7eJFpW5V4aBqImGmWwBgAmQ+gFGv7XnA+bqts1d7j3YXXEI7kftWLmG6BQCugeADKMA0BssWoGYUCvh0W/Wn9GZ/NgtSDNMtAHBtBB9AEU0NsZzupPcceEVpW/JbGheABH2WKkMBplsAYBIIPoBJat5YW3Cfl3DQr7f3bS7hyABgdqHgFJikpoaYXttzv3ZvWpVzeyjAfyMAmAr+agJT1NQQU2N9VJayzcTygxEAwMTY1RYowLRX31oXZfUKAEwCu9oCN6Dx2Vd17Gy/0rb04tn+Ug8HAOYcgg8gT1ffgPN1RZCN4QBguhF8AHmCvrE26o9vvbOEIwGAuYngA8jT0rhaNZGw9u9Yk9PnAwAwPSg4BQAAN4yCUwAAULYIPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcIPgAAgKcIPoBJauvs1T0HXlFbZ2+phwIAsxrBBzBJhzt61BdP6HBHT6mHAgCzGsEHMEnNG2tVEwmreWNtqYcCALPalIOPEydOaNu2bYpGo7IsS0ePHs25/5vf/KYsy8r52Lx583SNFyiJO/f+Qk8c7VZfPKF/+X/vl3o4ADCrTTn4GBoaUn19vZ577rmix2zevFkffPCB83HkyJEbGiRQaolkxvm6q2+ghCMBgNkvMNUHbNmyRVu2bJnwmFAopGXLll33oIByVldTVeohAMCsNiM1Hx0dHbrlllu0atUqNTc369KlS0WPHRkZ0eDgYM4HUG7271ijmkhY+3es0bFH7i31cABgVpty5uNaNm/erK985StasWKFenp69P3vf19btmzR66+/Lr/fP+74p59+Wq2trdM9DGBaNTXE1NQQK/UwAGBOsGzbtq/7wZalf/u3f9OOHTuKHvO///u/qq2t1X/+53/qgQceGHf/yMiIRkZGnO8HBwe1fPlyDQwMaNGiRdc7NAAA4KHBwUFVVVVN6vw940tt77jjDt1888167733Ct4fCoW0aNGinA8AADB3zXjw8fvf/16XLl3SrbfeOtMvBQAAZoEp13x88sknOVmM8+fP68yZM6qurlZ1dbVaW1v11a9+VcuWLVNPT4++973v6TOf+Yw2bdo0rQMHAACz05SDj1/96lf60pe+5Hz/6KOPSpIefvhhHT58WF1dXfrnf/5nxeNxRaNRPfTQQ9q3b59CodD0jRoAAMxaN1RwOhOmUrACAADKQ1kVnAIAALgRfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE8RfAAAAE9NOfg4ceKEtm3bpmg0KsuydPTo0Zz7bdvWk08+qVtvvVXhcFgPPvig3n333ekaLwAAmOWmHHwMDQ2pvr5ezz33XMH7f/zjH+vQoUN6/vnn9cYbb6iyslKbNm3S8PDwDQ8WAADMfoGpPmDLli3asmVLwfts29YzzzyjJ554Qtu3b5ck/exnP9PSpUt19OhRff3rX7+x0QIAgFlvWms+zp8/rwsXLujBBx90bquqqtKGDRv0+uuvT+dLAQCAWWrKmY+JXLhwQZK0dOnSnNuXLl3q3JdvZGREIyMjzveDg4PTOSQAAFBmSr7a5emnn1ZVVZXzsXz58lIPCQAAzKBpDT6WLVsmSbp48WLO7RcvXnTuy/fYY49pYGDA+Xj//fenc0gAAKDMTGvwsWLFCi1btkwvv/yyc9vg4KDeeOMNffGLXyz4mFAopEWLFuV8AACAuWvKNR+ffPKJ3nvvPef78+fP68yZM6qurtZtt92m7373u9q/f78++9nPasWKFdq7d6+i0ah27NgxneMGAACz1JSDj1/96lf60pe+5Hz/6KOPSpIefvhh/fSnP9X3vvc9DQ0N6a/+6q8Uj8d177336pe//KUqKiqmb9QAAGDWsmzbtks9CLfBwUFVVVVpYGCAKRgAAGaJqZy/S77aBbNHW2ev7jnwito6e0s9FADALEbwgUl7qv0t9cUTeuJot1bsaScIAQBcF4IPTNpwMuN8bSsbjAAAMFUEH5i0bfXRnO8TyYzuOfCKdh05zXQMAGDSKDjFlNy+p33cbX5LSttSJBxUZSig5o21amqIlWB0AIBSoeAUMybos3K+tyRtrYuqJhLWSCqtvnhCe492kwUBABRF8IFJa+vsVcCfG3xUBH06tHOdXttzv0IBv6RsPcjhjp4SjBAAMBtM6662mLvaOnu192i3zBydpWyQYQIOSdq9aZUOHj8nSVofW6w79/5CidEi1cb6qA7tXOftoAEAZYnMByblcEdPTuCxrT471bJ70yqn/8e//L/39fFwUvetXKJTvVecwEOSjp3tL8m4AQDlh8wHJqV5Y62T1bit+lNq7+rX6miVDnf06PLQiBLJjPriCUnSi2f7VRUOlnK4AIAyRvCBSTGrV55qf1tdfQOS5HzOF/BZiieSObdFwvyqAQCymHbBpB08fk6JZHrc7eGgT5FwUJFwUPt3rBlXlCpJHw+nvBgiAGAW4HIU1yXos7Tl87fqVO+VcX09skFKxjkulbG1tS5a7KkAAPMMmQ9M2u5Nq2SSGhnb1t0rqtW8sVaHO3py+nrs3rRKNZGwgj5LyYztLMcFAEAi+MAUNDXE1Lp9jdPRtOWFbh08fk598UROX4+mhpiaN9Yqmcmuj0kkM9p15HSphg0AKDMEH5iS/ABEkmoiYTVvrM05Lr/JWHsXS20BAFnUfGBK2jp71XrsTSfwuG/lknFTKm2dvRoaSY3u9eJXX3xYq6NVJRgtAKAckfmYxUxzLy/3UTnc0eNMp0jSiXc+KniMWWp7YWBYknRp6Ko3AwQAlD2Cj1nsyRe61RdP6Imj3brnwMuevGb+9EqxY7IbzWWUtrMdUSfzOADA/EDwMYu5EhDqiw978ppNDTE11kdlKdvfY/emVQWPyW40l/31qgj6xq2IAQDMXwQfs1hNpKLg1zPt0M512rdjjaorQ5KKT/+YJbehgH/cihgAwPxl2bZtX/sw7wwODqqqqkoDAwNatGhRqYeDIu458Ir64gnVRMKS5Hz92p77JWUDksMdPc50S+uxN5XM2KqrqdKxR+4t2bgBADNjKudvMh+4LqauY31ssbOyxV3XYfp/HDx+Tk0NMadItdh+MACA+YPgA9fF1HWc6r2ieCKpylAgp8V6Pp+V+xkAMH8RfOCGmAxI/mqW+1YukSVpJJVWW2evFlUEJcn5DACYv2gyhhvS1BArmPE41XtFtrKt1Q939Gj3plU5NSAAgPmL4AMzonljrQ4eP+d8XSxIAQDMP6x2AQAAN2wq528yH/BEW2evkwnZvWkVWRAAmMcIPjCjTL+PoZGUs9+LWX4LAJifCD4wY9o6e9XyQrezvwsAABLBB2bQ4Y4epUcrimxJkXBQlaEAK14AYJ4j+MCMad5Yq8MdPVofW6xTvVecVS8AgPmNJmOYMaYLqiRdGEjo5PnLJR4RAKAcEHxgSvJ3sC22o63bsbP9StvZzwAAEHxgSg539KgvntDhjp6C3883u46cVu1j7dp15PSkAjEAAMEHpih/L5f1scXyW9nPxTTWR+W3sp/n0gm6rbM3J6vzxNFu9cUTaj325pz5GQFgJtDhFDfkngOvqC+eUE0k7NR3TOfx5WrXkdPXnEayJJ0/sNWbAQFAiU3l/E3mAzek2K6203V8uWrvGgs8GuujCgf9446xlQ1SyIIAQC4yHxjHdCX1emls47OvqqtvQHU1VTr2yL2eve712HXktNq7+rW1LqpDO9dJkta2vuR0cS1k/441LDUGMGext8s819bZq6fa31YimVY46NeX71qqE+98JEm6b+WSa/bccBeRXs/J8nqDl66+gZzP5ezQznVO0GF+Xvd7e/L85XHTMi0vdEsSAQiAeY9plznGtDRPJNOSpEQyrWNn+xVPJBVPJHXsbL/64gm1vNBddCrgRqdGCq2AcReaFis6raupyvlcrvLHb37eU71X9Nqe+9XUENOp3ivjHpe2NW9XBQGAG5mPOebg8XNOS/OJpO3iV+JNDbEpXZ3nT5eYzqbu4CU/IDEBkPv1y32qxcjPDBX6edfHFqsvnsh5XCQcnPW1LgAwHch8zHLuPhOSNJJKX/MxQZ8lvzXxlfhkl8S2dfbmTJc0PvuqJGloJKWDx885j2/eWKtIOKihkZSzPHe2ZgLyM0Omk6s7YDPTXG6VoQBTLgAggo9Zzd1nwr36wi0cHP9PXBkKqHX7mgmnVibTPMxM8bh19Q3ocEePM81jApDDHT0aSaUVTyR14p2Prvn6+a9zrSmbmVLo9QoFG/mPGXAVngZ9FlkPAHAh+JiFzAnxqfa3nNsWBHyjJ8jczeu/fNeynFssSbdVf8rZ8O1wR0/BE7k7U1HsRO/etdaoiVTknGRNANIXT2g4mXFuNyfwk+cva8Wedt2595cTvo4JhMxzHTx+rvCbc4PyA52WF7qn1MHVPMaW5LeyK1ze/dGf60zLQ2Q9AGAUNR8lZrICQyNJxROp0ZP3ZyZcLXLw+LlxSzoTyYwOd/QoFPA5xaaSdKr3ivbtWKODx89pIJGUrbHVJKYmwd0q3QQO5rZ4Ill01Yu71sEECJKlpoZYzhjjiaQsSdFIhS4MDOu+lUuc52jv6petbGFs/jjMazZvrNXB4+c0NJLS0EjKec7b97RP+7Lc/IxP2s4GEYWyFu5GY4312SW3JiDzW1LrdpbWAkAh9PkoIXOVnJ89iISDzom7sT6qu1dU5wQGTxzNneoIB32qrgw59x88fk4jqbRCAb+z/HNoJFWwB4UlqSLo03AyI1tSTSQsKRuYRMJBVYYCWh9brFO9V5zPhYIiE0S5jz3xzkdOwGNey7yG6W6afwI/1XulYAdU0xnVPEf++7V706ppOdG7lwlL4wMhI7/Dqd+Sep7eWrIeKQBQalM5fxN8lJA5oZriS8MdfFiSfK77w0GfEqPTF0GfpVsWVTgnukInPvMa7kDixDsfaSSVUSiQnXVzv1ZF0KeRVEYZW05WIX+cE7VGN422IuGgzrQ8pLbOXu09mp2GCAd9CgWynUDdwcKde3+pRDKtoN9S5YLAuPul3ODmF7/+QMnM+F9bk33wQu1j7Tn/Zl6+NgCUI5qMlTlzIr2pcoEuDCS0tS6acxXtzlDYyg1MEsmMc/VfGQrkBAGFmoO5p0YKXYm7r+ArXIGNJL3Zn52eWR9brAsDCa2OVunS0NWihZPuQsuRVEb3HHhFzRtrtW/HmnHTM+4pFjNNlEzbiieSqomEJ1z+W1ukuPbY2X7dvaI657GFOpFOB/e/GYEHAEwNmY8SyG/DXRMJj+sJMRGTQSiWHZhKyt9kNaRsxkWShq6mlErb2jZ6UnVnPiaqY3A/l8nQ5GdJ2jp7nWLRkVRaiWRGPksyiQzzGpJypnHcP5PpK1ITqdDQSHrce2ler62z15miYpM3AJhZbCxXxvKXYUqmvmLySahEMlOwZ8S1loAWYla1RMJB3bdyiSpDAbVsW619O9boVO8VNT77qhNQXKsvR/PGWvlHl9aEAv6Cq2WaGmKqDAUUTySd1S92XuDR1BBzMiSmI2vrsTed57g0dHX0K0tnWh7S/h1rnJ9hfWyxs1rFvSLGlnLG4fWSXQDAGDIfHipWYOpmSbJcmQDDZ2VP0oHRuojrLbCcKDvi3u5e0rhsjCVp346xrMS1Ck/bu/oL1ogUKk412Q3z3OY29xh+O5q5mMzPEAkHc4pdpdzgxv2zFqtfAQBMHpmPMjSZwEOSttVHdWtV9uTvt7L1BJFwUBk7e/WeTNv6eDipk+cvF32dia7oJ2oe5u7cab5277NiT+I5TPblVO+VostUzTGHdq7L+ezOeJh9Ugrt9zJRhseM2z1ew525udH9awAA14+CU49Mds+VY2f71Vgf1dBISiOpdME23Wm7cHGlND4wmKhnRltnb87j8/d0MV+7a1T2Hu1WwGfJUrYQtZC2zl4NjaQmXALrnhZxH5O/T0qhHh7Xqm1x9wMxIuGAKkNjXUanun8NAGD6kPnwyGT2XDH+462LiieSSiQziieSGhhdurp/xxo11ked4/IzD+6TfqGVJVJuzcVk93XZvWmVU4xqS0pmbNkqvH+JGVc8kZxwL5On2t9yWrA/cbRba1tfcoKhQlkN95gmyryY5mZmKa7fynZdjSdSuqlyAQEHAJQBgg/PWEXviYSDCvrG7h9O5gYqtrLLb0+ev6xDO9epsT4qvzU+82BO+ubr9bHFBacWrjXlkH9yb2qI6UzLQ04AYoyk0gWneMzGccUyI9mfMZPzfX4wZIKNXUdOa23rS9p7tNtpq+4OsCYSDvq0sCKovviwpLHOru7np+AUALxH8OER09ArX9CfXbHR0rhaNZGw9u9Yo231UVmSwkF/zsZwL57t1z0HXtGJdz5S2s62TnefRN31Du66iamuiikWnOzetGrcRnWFMhBmfC+e7S96cjc/o1kdEw76cl7PBEDtXf2KjxaO+i05m9OZn0PKDSR2b1rlvI/VlSHFE0knsKuJVEwqewIAmFkEHx4pdOKW5HT0NAGBNLYfy9v7NuvxrXc5J8+Az3JWf5jgIP8kmq0VydzQLqrFgpOmhpje3rfFCXCGkxO/jq3iS3MP7Vyn8we2atlocW0o4M/Z5M4EQFvros4y2tbta5QanU6JJ5LO5m/urIh77OY5WhpXa/+ONbowMDzl7AkAYPqx1NZj+ctM84sm3UtATXBh9mUxLdLdj3EXX45t7pbNEmytixbdi+VGfwazcqfQUtVixaQTvR+Xh0aUGA1mzrQ8VPT42/e0O19HwkF9PJx0CnmDPkstjatzXltSzqZ6boXez2LjY68WAJgYe7vMYu6T3VPtbymRzIzbw2Wixxba8TYc9OntfVtmbJw3elJ27/9yreDjzr2/cFrAh/PawVuSqlz74hTrV2KOlcZvdJePfiAAMDkEH3PEij3tztW6yYRM5kSf3759Mq3FJ7ub63Qyr2kyO35LWlgRUDyRUk2kQq/teWDcse7GZC+e7ZctOe3ZayIVTnGpJO3fsUaHO95TX3xYPkv6P3VRnXjnIw1dTSnpWvdcV5Pds+amygV6s38gZx8YMh8AMDkEH3OE2RRtQcCvRDI96atv98Z1v+4bUEXQr8e33ilJRQOMQt1NZ/Jq3z11457+MHuxSHLas6cytgJ+S8m0rXDQp+rKkBOwuDMYbqbHiPv59o9ucNcfT4w73s1vST1Psw8MAEwFwccccyNX3+5N4XyWpWTGdnbFlcamHLzMfLgDj/zN6u458HJO9qKYoN9SJmPL57NyshiWsoW5qYyds0uvz5IWVQRzMkLuDe3cfJb0wwk20AMAjEfwAUljNSCFii2N/Tu8P8kW2iW3WOFsPnfg5P7eZESaN9bmtLEvdrz5uiKvbsQIB/16e9/mG/xJAWD+YG8XSBprOlYVDjrTKW6RcCBneatXzBJYd8bDvWTY3O/qu+ZwBxI1kbC21UdVEwnrs7cs1IWBhE6ev6zV0bF9YCqCvpz2blWjy3ZNEHI1NT7wkKREMk0DMgCYIWQ+5rD86ZpdR07r2Nl+5/7c6ZfcAk+vFZpachfcNtZHnQJTKZuZeHzrnU7WxNR2WJKikXBO5iToy043mWmWcNCnL9+1TKd6r0iyJzXNU4oMEQDMJmQ+IGl8s7BDO9c5LdItSZYrJdAXH/a83bi7hbpp/OW2bbSNfGN9dvVJhatJWyKZztk8zwj4LDVvrM3Jdph9Xkx9RyKZ0X+8dUGSxgUexZrg7z3aTSYEAKYJwcc8Y9qP79uxRvk5r754Qk8c7dauI6c92fskv4V6PJFUywtjJ/lDO9epdfsap418KOB3HhsO+p2deXO6lFrZ5/18TZUsqeDUjZQNQPLrSsJBn/btWDNuDxtp4m6tAICpIfiYZ9zZkLE9ZHw5u+UeO9vvtCyfyRNu88ZaRcJBLXDte5O2sx1JTfBz8Pg5Zxy7N61y9oK5OrrHy+GOHjU1xJyAIZW21RdP6M3+AdmaaDs/OVkVc0wo4Hc20Ss2XgDAjZv24OMHP/iBLMvK+fjc5z433S+DaWD2V3l73xbdvaI65z6zkdtMnnCbGmKqDAXGrTYZSaVzVryYfh9StmW835JWR6uc282GcpFwUBVBnyLhoLbWZQtRV0ezGZCgz1IkHFTdaEYkHPSpdfsaHdq5Tvt2rFFNJOy0Y5eyjcfcwkEfNR8AME1mJPOxevVqffDBB87Hq6++OhMvg2lUKMOxIODTyfOXZ3T6xaxscWcoEsmMbqpcoHDQp4FE0tnJ9nBHj071XlHalt7sH5CknOyHCWQ+Hh7r5fG7y3+UrWzdRzyR1NsXBmUruyme4d7Uz/ysxx65V/tHp2Ai4aAe33rXjPz8ADAfzUjwEQgEtGzZMufj5ptvnomXwTTKL9KUskHAi2f7c2pBrqWts1drW1/S2taXJgxYzLSKJL225/5xtRZdfQNKJDNOoGCyHOtji+W35PTxcGc/1scWS8re197V72ROcopPRx9oS84GdEb+DsFmCuZMy0NkPQBgGs1I8PHuu+8qGo3qjjvu0De+8Q397ne/K3rsyMiIBgcHcz7gvaaGmDP90Fg/to19RXCsyPPY2f5rZkFMbxGTkSjG1HI81f6WE4QUq7X4fE2VKkMBxRNJneq9otbtY9Mk5naTFZHGdvQNB/0aSCT1+ZoqhUf7fZhpF8NdWLs+tliWpA8Hh3OCJy+KbwFgPpn2Ph+/+MUv9Mknn2jVqlX64IMP1Nraqr6+PnV3d2vhwoXjjv/BD36g1tbWcbfT56M8uHtoGO49X/L7c5iuqiOptEIBv+5buUQn3vlI0tgW94c7enR56KoSybTTa8Q8Z+Ozr6qrb0A1kQpdGBhW2h7bVM+9sZz79Yq1ha99rN1p4e7eq6VQN9VCO+CaMbGzLQBcW1m1V4/H44rFYvrJT36ib3/72+PuHxkZ0cjIiPP94OCgli9fTvBRRvJ3lHU3Ait2Yna3UHe3Ojdb3puv71u5JOc5zeMkOQGIe5dZc797MzppbApl96ZVzthMIFNXU6Vjj9w74c/mfh4TON1W/Sn9um9AAZ+lylAg57kBALmmEnwEZnowkUhEK1eu1HvvvVfw/lAopFAoNNPDwA1oaogVPema7EH+qhh3psJ0JjVRrglIKkMBHdq5zgkAzONMpsU0APuPty7ongOvqHljrfO8QyMp9cUTOnj8nD4eTjoBjik+laRLQ1dzPk/0s5kxuAOM2sfac4pVAQDTY8b7fHzyySfq6enRrbfeOtMvhRLI76Kaf7tZympqSHZvWuXUbLinSkyhZ1NDTI2jnU2Dox3ChkcbgrW8kA1KXttzv9MsTcoGMpayxafmOds6e/XhYDZ4ualyQcGxu2s53GMwt7v3iDHjBADcuGnPfOzevVvbtm1TLBZTf3+/Wlpa5Pf7tXPnzul+KcwShTIn7u/XxxbrwkDCWa1yaOe6nIzI+thitXf1K21LLS906+T5y85UjZRb52Ec7uhx2qqbZbn58jezK7Srbk2kQn3xYflmuOcJAMwn0x58/P73v9fOnTt16dIlLVmyRPfee686Ozu1ZMmS6X4pzBGmd4dZrWK4g5a7V1Sr5YVuZxlt2s4GD/lZF3fBa9BnKZWxtbAioNv3tI+r/WjeWJuzp4y7ZqVQQAMAmB7saouSMsGClLsaptCJv1jhq7to9ODxc059himCvX1Pu/Mc+3esyXl+VrIAwPQoq4JTYCKmL0hNJJwTDJj6DncAUqzwNb85mGGmcYod29QQK1owCwCYOWwsh5Iy7dXNyb95Y62zGqZQgWd+w6+2zl4NjaScYtPdm1Y5TcRMfxGzeVzQZ+ny0EhOYWqxglkAwMwh84GSys9mmK+LZSNM5sJkRtyZk5PnL6u9q18Bv+W0UZeyBayneq+oL55QMmOrujJEsAEAJUTmA2VnomxEfmbEZE7Wxxbr2NlsIWoybTvt102mxL3c1kzH0DYdAEqD4ANlzx0kNDXEcvqEmEAlf6WMuc9kStzLbc10jNlfJn+DOQDAzCL4QNkrtNtsfmakeWOtIuGgU+/R8kK32jp7nczI1rqoc2w8kZzUDr0AgJlB8IGyMNEUyPrYYvmtwqtXzGNNb4+K0d1rzbSMu9Oq27Gz/U6XVLPEFwDgDQpOURbyl8C6FWtC5n5s/t4r7hUtxUy0Zw0AYOaQ+cCMm0xhZ/6S28neZ+6PhIM5t1WGAuMCi8b6samXcNA/lR8BADCN6HCKGWcah1mSqkY3l7uejIO7k2mhx69tfcnJgDTWR3OmWop1RwUATI+pnL/JfGDGmYyFrWyx5/XuDlusk6mxe9Mq+UcrTk+881FOtsU81qx0AQCUDsEHZlxTQ8xZhSJd/+6w7p4ehaZx3MtwJeUEKuax+bcDALzHtAs8sevIabV39WtrXXTcypOpMtM4QZ+lZMYet1utlDtFI6ng10y7AMD0mcr5m+ADs44JLPriCee23x7YWvR4dq4FgJlHzQfK2o22NTe9Oyb7Wu6N5wAApUfwAc9dq3B0supqqnI+F3Lw+DlnBQzTLABQHmgyBs81b6wtumvtVOTXeUxkJJXWPQdeodYDAMoANR+YU0yrdUlO2/TDHT0aGkkpnkhS9wEAM4SaD8xbptW66Sdi6kPMPi7UfQBA6THtgjmleWOtk/lwBxrs4wIA5YNpFwAAcMOYdgFG3eiyXgDwynz6e8W0C+Y097Jepl0AlKN7Drysvviw8/18+HtF5gNzmtnThUJTAOWmrbNXd+79RU7gIV3//lezCZkPzGkUmgLwUqF9pdbHFutU75VxfYYOd/QokcyUaqglRcEp5gX3HwSCEQAzZW3rS4onks7Gl26mz5DpRzSSyiiVzow7LhIO6kzLQ14Oe1pQcIqyVaqCKlP7cfD4uXlT0AXAW22dvc52DvkBhSRdHrrqXAjFE0klkmndsqhCkXDQ66GWHMEHPPVU+1vqiyfUeuzNokFAW2ev1ra+pLWtL01bkLA+tliWpHgiWXBfmflUZQ5gZlxrv6pEMq29R7u1PrZYkXBQ4aBfQyMp3bdySc5xt1V/as7/PSL4gKeGR+c3kxlbffGEWl7oHvcfzGwGZ7qU3khgYAKZY2f75b4O+XBwOOf5pmuzu9nAFLmt2NOuXUdOl3o4wJxRrFDUndmwJZ145yPdt3KJEsm04omkXjzbn7NB5pv9A3P+7xHBBzy1rT6a833aLn61YGlsE7r8KZOJAhL3fe5dbd2SGVtPtb+l2seyJ+D5tCrGFLnZko6d7dftBCHAtGhqiKmxPiq/ld1tOxIOKhIOavemVQoHx063A6MBh2FLujR0Vft3rFFNJKytddE5//eIglPMuPxiz11HTqu9q1+ro1W6NHR1XBFo/vHme7M5XCQc1MfDSaXt7BVFZSiQc2zLC91K29niLvOYazH/0edDMWpbZ6+ean9rXJV90Gfp3R/9eYlGBcwt9xx4RX3xhFNkagpRDVOQ6rOkRRXBnI0wZ+vfoqmcvwk+MKPczXPCQZ8e33pXTnBgdpjN34220H+8XUdO60XX9InfkhYEfM5JtK6mSu9++IkSybQkaf+ONZKkvUe7nceEg36FAj7t3rRKT7W/7Rxrnm9rXbTgkri56LPf//cC1fgVem3PAyUaETB3mIusrXVRHdq5zrmIMstuzYWR35Jat69RU0NsXMAy21bpTeX8TZ8PzCh385zhZEaHO3qUtrMnendK0VR/S1LLC93O7SYguW/lkpzAQ8oGHm5dfQOyRr82nw939GhbffGAwvwxaO/qV9qW83k+dBh890d/rsZnX1VX34BzW198WLfvaZckNdZn/2gCuLb8QOFU7xWlbelU7xVJuT2HzMWWJeX8vTEbYw6NpHIutvYe7XaeY66g5gMzqiZS4Xy9rT7q1FaYSN9o3ljrFGWl7WzQ4S48be/KDTwkKZHMKBTwK+jLhhp1NVXaNjrfuq0+6tSKnOq9otf23D/uP25TQ0yv7blfh3auU+v2sbnWSDiooZHUnK40N449cq98VpH7XHPSACaWX7RerI7MTA3HE0lVhYM5xzQ1xFQZCjhFqOZvni3pqfa3PfxpZh7TLphxU0kd5s+LGo31UZ145yMNJJKylc1sVI0WchV7zutNWZoxWJL27Vgzp642CsnWgOROQRm/PbC1BCMCZp/J/r0xUyvF/oaZ5/lwcDhnWtSSdL7M/z9S84GyYv6zuec2C9l15HTRq+3G+qjuXlGdMw1jplJOnr/szK3evaL6hudI3QFQfkHrXJdfo/P2vi0lHhEw+7nrPU6885FzezyRzKl9c8u/EKurqdKxR+71ZLzXi+ADZaWts1dPjM5ZhoN+vb1v84THFOK3JJ81Vh1u5kprImH1xxNONiQaCecUbF3veE2QI038BwIArsV9AWb+bpk2AsUubPJX+c2Gv0G0V0fZMCl9I5FMa23rS/rs4/+u2/e06869vxx3speymY6aSNipGVkdrXJSkBk7+x/Y9AGpGF0/XxH0TUu/jqaGmM60PKQzLQ9p96ZVc369PYDpld+HyPxdmqh/R/5jTE3afSuXyG9luzTPJax2wYzKNrTKrSVwpxITyXTBnhNmlcU9B16RlG3AEwkHFE+knGOs0ULJUMCvUMDvzJ1O5/SI+/nMlATLUQFMxF186v4bYqZe3BkNc0z+Y4z8VTNzBZkPzKjmjbUKB/3O93U1VTmd/qSxlutGo6sLqjuT8fFwKue4jD22RLcyFJjxmgxTC9EXH9Y9B16e0dcCMPs0Pvuqbt/Trv7RglJ3tsIEF+1d/eqLJyQpJwuSn7U1mZD1scVzMvtK5gMzqlAmwsx/SmOBxotn+1Ux2oTMfbz78SfPX84pSK2rqdL//dPlzrzpTPNZ2YBHygYgbZ2986IIFcDE8vvlmEJKd7bC1HiYJmOm+PTg8XM6ef7yuF5EJliRVPa1HteDglN47ka69n328X9XMm0r6Lf07lPetgJv6+zVky90OwHIbCgAAzDzTGM+w90yvdjfOPdFmBEJB3Wm5SFJGtcRdTasuGO1C+Ysd+AilWYfhNnW8hjAzHIHH+4AYiKFVvgV2l/JLLmd7POWEqtdMGeZCvD8Aq1SjQEAGuujspTtjWM2iLseqUzxXMBIKl10J+/ZiJoPzFrudfIAUCqHdq5zGhxOxLQVMLtyu1nKbguRP91iGioOjaQKroaZrZh2AQDgBk1mR9pCdR6G2dvKbCFhmL2rkplsrVvLttVlG3ww7QIAgIfyl8oWmhYu1CisrqZKNZGwpGwPJFvZjs5GMmM7DRaTaVtNDbFxDclmI4IPAACukwkEJOXUghXqtlyoUdib/QNq3lir3ZtWKRIOKhIOqnX7Gqc/UtAViZgvTWCz92i31ra+NCuDEKZdAAC4TvnTLRNp6+zV3qPdspUtTr2ayjh7vZjHFlpie/D4OcUTSYWDPlVXhjQ0kszp9hwO+nU1ldbWuqjTHboUmHYBAMADJsOxPrb4mlMhTQ0x7duxRjWRsL581zItrAgqHPRraCTlPM5kNdyBx9DVlEz+oy+eyAs8fBpOppW2pfauwruClyMyH8AMoBcIMH+0dfaq5YVupe3s1Ejr9jXX/H9vMiaWxjqiNtZHnVUzZjWfu0A1Eg6qMhTQhYGEs1rGb2U33nyzf0Cro1V698OPlUhmFC7QMXqmkfkASqxUPUiAUjM1ELuOnJ71RZGT4Q48pOyO208c7Vbjs69O+PNn973y5axsOXa2XwePn3MuWpo31joZD0vS7k2r1LyxVgsCY6futC397vIftbUuqq6+AWeTzkQyo71Hu8v2/SfzAcwAMh+Yi/I7DB88fk4jqXTOrtKmI6e5op9r2xCYXh0jqYxCo0GA2am7JlLhbEBpmGyFu4bj5PnLau/qH9frwzDZjF/3DSjgs1QZCjjvb6HlukGfpYxtF3w+L99/2qsDAKadu7hS0rgpgTMtDznBhymOnCsBuAk6TKBRTLYGI6PP11Tp0tBVDY2kFE8k5bdUNNhwc0/DGKYHiCTdt3KJTrzzkUZSaSfLkc/dG6SupkrHHrn32i88DZh2AQBMuw8Hs1f1ffGEbqpckHPfQCKpXUdOa2gkWwyZSGb04cfDOnj83JyYgplM4CFlf+5oJKxjj9yr1/bcr92bVqkmEtbWuug1H9tYH9W+HWty+nxI2cyK+Wjv6td9K5foaqpw4CFJsuT0Bvl130BZvvdkPoBpwDQL5oP83Vsny301HwkX3+3VZBckTbgjbCmYjI5hSYpGKtQfH3Z+tmtlexqffVVdfQOSsu/D0EjKCRJM5mjXkdN68Wy/LEtyb/VSKCMyGeGgz8mQzHQRKpkPwGNPtb+tvnhCTxzt1u172rXryGnnvrnQjRCYiM8qfHvQb40rqownkjp4/FzO/4m2zl7d8Vi7njja7Vzhl1uxdv6GcbakCwPZwCMSDqomEtbjW+9yVqm4/7+bvwG/u/xHSXKKSCtD2e3V/NbY8794tl+2soFHJBxUOOhTJBzUtvprZ07yRcJBffmuZc73iWSmbN5XNpa7BhOJuyPUxvrSNnJB+RlOpnO+b+/qd35H3CtfTp6/rGNns2vxS7EUDrgRkXCw4NTDoorxt5tCy6GRlHPlHRwtnpSUs0na4Y6ecVf55bZhpPl/6m4Alv/Z3N8XT6jlhW7nceY2S2P1G/FE0qnNWB2tcp6/wpWpMO/pl+9aVrA76kSs0cef6r2iupoqdfUNKOiznH4kpc7SkvkooK2zV2tbX9La1peceb6k639GsUYua1uP6/Y97VrbetyroaJM5F+VuOd3mzfWKhIO6vLQiBN4SNmrkNZjb3o2RuB6mb+Jpp7DLVgk7TGSSqsvntDA6MoXSbplUYXOtDzk1EGYACM/0Ni349p9MkqhqSGm1/bcr0M71zmfmzfWqr2r3wmmmjfWOsWlJstgbjNnkZFUWpbG6jK6+gacNulfvmuZ/FZ2zxfj2Nn+cTU2hqXcY42qcEB+K7ufjMm4VIYCOvHOR+qLJ5zprVIh85HH3f5WGkuPua2Ojv+HbuvsdbrOubvPYX44tHPdhNmwYoVq7qAWKFeHO3oK/g77RgsbC90XCvid9uHmit8EGU0NsZzgoqkhpoPHf6N4IqVIOFCWgUcxhzt6nOZi7myCe0lyU0NMJ89f1otn+3Peq3BeluPg8XP6eDiptC1dGrqqxvqoc8FiakXM41IZW8m0rYqgX5eGro4blzkPtXf1a0HA79w+kkrnfC4Vgg+X/MBDKlzg82b/gHYdOe0sdwoF/Dn/kJEwbyvGlMscK3C9TJtv09vCLPecaPWHqWF4YvRv6kgqPWFQcaZl03QP2xOmxsMEHsWKz0/1Xsm5qK0aLbyVsjVjiWTaeT9NxuJU75WcAEQam/YfK4C1NTSSUtBvKZW2c85ZlrIZmFDAp+rKBbqpcoETxIRcAUkpcJZ0OXj83ITVxL7R6uO0rXHpc5MhMRXLgNG8sVZPtb9VcE1+sUI9oFyYk6l79ck9B14ZF3hYGqtXiISDzrHmgm64SE+K2c6d6TCfzbSGOwgxAZw0tpLHrGzJP+/4LMu5vT+eUNBnOVnSU71X1NbZ60zdpDK2Eq6mboVlAxR3X5b8AlqvEXyMauvs1cA11nBPlCHfVh91utcBbibFnN8d0v01UI7cS0Ofan8rZ7t4cyK9b+WSnL99+b/X2+qjau/qn1Sfi9nKXVS+PrZYFwYSGkllFE8kcoISc8I/ePycDh4/p4FEctwUv63c6VjzvSngNZkWczGTHO1cVuHaJdf9WGl8KUBjfbTkU1vzus9HW2dvwSvSSDiokVRGieT4OTGT/cg30dp1AJht7jnwck6rcEvS+QNbSzegMua+sDCBSH6wUKgzbNBnKZWxFfBZStt2wXNLOOiXZCsU8DuB3vrYYp145yMneDHTOPetXJLTtt1kTNy1JUG/pZZtq2fkXDWV8/e8zXzkbwbk5l4Cla9Y9sOsSzdFRRVBvx7feifBCIBZp62zd9weJdfTZ2K+yC+gzc9wro8tljS2queJo9lluAG/T+/+aHPB/VrMNEp15QKnRbt7imb3plXaO/o8trLnoBPvfKStdWM1IiaD4r7ATqZtZ4lzKc3bpbZPtb81YZ99k+bav2ONGuujslR8SZk0ti69vSv7y5FIpik0BDAr5f/tqqupordRAYUaCJr6jsMdPTp4/Jz64glnWsq8r6bXx3AyrbbOXic4MRrro9pWH3UKT42KoF81kbDWxxaPWxwhjQUg15L/eqUw76Zdim0ONNFuhKZgKL+9rpSdhllUERxXQETmA8BsVc5tzsuJyVjk71xrMhXhoF/DybQCroLRcNCnUMDvTJm4p2L8ltS6fU3O7rV+K9s3yEy3/MdbF4puKGfklwe4C1aN/TPQS4VdbSdw595fFPyHc1cKF/rleG3P/QWDD78l9TzNPCgAzDem1sMEG25Bv6WAz7pmoBAO+pwOpuZCd9eR0zkrKt3dYiezuZ3b/h1rCl5wz8TKTPZ2KWLXkdNFfxFM4OG3pMe33uXc5m7zW2hp0lyu4AYAFGc6nu7etGrcTrTJtD2p5fWJZEaneq/otT33O5mI/C7aplvsVJlVLfetXCK/lc3wG/FEsqT7Tc1Y8PHcc8/p9ttvV0VFhTZs2KCTJ0/O1EtNWrG26Nlq4iyfz9JT7W85UaKt7LKots5eNTXEnPqPcNCv/TvWMA8KAPNcU0NMrdvXOLUc+Rrro4qEg4qEgyo013BT5QKnfmTXkdNaEMg9NZveKbs3rVLjJAt/I+Gg7l5RrbWtL+nFs2YFjJXTtbuUdYkzMu3y85//XH/5l3+p559/Xhs2bNAzzzyjf/3Xf9W5c+d0yy23TPjYmZx2KdbQZTJoHgYAmEihVSvhoE+SpUQyrXDQr8/e8v/ltEqXshn3haOb85kSgJpIOGeaxeyFc7ijRx8MJCbsO2WW3ro3RJWytR933rpIv+4bmJG6xJLXfGzYsEF/+qd/qn/8x3+UJGUyGS1fvlyPPPKI9uzZM+FjvejzYYqp3A1eCgkH/U6vD4IPAMBE3LUaQZ+lgN9SKOAvWKfhrjOsq6nSr/sGZCt73qmuXOBM97sLf02/EHffDjeTeZno3DaT57KS1nxcvXpVp06d0oMPPjj2Ij6fHnzwQb3++uvjjh8ZGdHg4GDOx0xraojpTMtD2rdjTdH21vt3rNHb+zZr/441qomES96KFgBQ3swy13DQr1sWVTiBR6HzjDs46BoNPKTsPiym8LSpIZbTFXV9bLEi4WDRVvX3rVxyzYvqcjHtTcb+8Ic/KJ1Oa+nSpTm3L126VL/5zW/GHf/000+rtbV1uocxKe7GMO42wu59CfKbxwAA4GZWvZgNRhPJsQLRcNA3pX1t3A0r27v65bNy93WpDAVypmeMcNCXs3ld0G8pk7Hl81lKpu1xGZVSK/lql8cee0wDAwPOx/vvv1+ScRx75F6yHACAKTPNxFJpW/lJjkQyo8Bo6iMSDqgmEh6tAynu8tCIUyRqAg+z8rJ5Y61qImFV5D1HKOBX88Zap7C1ZdtqtW5fo8oFAUXCQT2+9c6cFTWlNu2Zj5tvvll+v18XL17Muf3ixYtatmzZuONDoZBCodB0D+O6kOUAAEzFriOnnZqO/EZehrl9IJHSmZZNTqZkfWxxTj8PI5HMOI3BsstjrXEb9x08fk6JZEZ+K7vT+m3Vnxp3DjO7D9dEwmV3bpv2zMeCBQu0fv16vfzyy85tmUxGL7/8sr74xS9O98sBAFAyxVo4FBLwWc6Ch6GRlO5eUZ1zvzsjEnAah1h6bc/9kqSWF7qd3XNN7w6zTcib/bkraCQ5WZJymWpxm5GN5R599FE9/PDD+sIXvqC7775bzzzzjIaGhvStb31rJl4OAICScG/kZtTVVOnN/oFx+4dVhgI63NHjZEoOd/Q4GY6gz1J1ZcipFQkF/KquDGl9bLHWtr7kFJL6LTlLbtN2NmC5msoUbHhZztn8GQk+vva1r+mjjz7Sk08+qQsXLmjt2rX65S9/Oa4IFQCA2ezQznU6tHNdzk7pl4auqnX7mnG72zZvrNXJ85fVH0+oIugbd5+Uu7RW0rjd11dHq5yAwjyuXAOMicy7vV0AAJgJppajUECQvw+MyVisjlbp0tDVgo8xTcvcK1ssSft2rCnLwKPkTcZuBMEHAGCuMZuaBn2WbllUof54ImeprNnAVBprhDk4nFTGzt113SyZ7Ysnch5TDthYDgCAMmJ6faQytl7bc7+21Uflt7L1Ie6iUDN9E08knRbqJvDwW9KX71qqoZGUIuFgWRaSThbBBwAAM8wEG9tGN4Y7tHOdep7eqv/7p8s1NJLSwePn1Pjsq3riaPe4QlVja11U7V39iieSqgwFymrKZapmpOAUAACMMYWp+dyrX+KuDeca66M61XslZ6O6U71XlLbHVrzMZmQ+AAAoEXdX0rqaKvmtbOBxaOc6vbbnfjWOZkwa66NO347W7WtmddZDouAUAABMAwpOAQBA2SL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAniL4AAAAnmJjOZSNts5eHTx+TpK0e9OqWb93AQCgMDIfKBtmd8d4Iqmn2t8q9XAAADOE4ANlw71FdCKZUeOzr5ZwNACAmULwgbKRP83S1TdQopEAAGYSwQfKSmN91Pk66LNKOBIAwEyh4BRl5dDOdbp7RbUOd/TkTMMAAOYOMh8oO00NMb22535J0j0HXlFbZ2+JRwQAmE4EHyhbhzt61BdP6HBHT6mHAgCYRgQfKFvNG2tVEwkz/QIAcwzBB8pWU0NM62OLtfdot+7c+wumXwBgjiD4QFlr7+qXrWzfjydf6C71cAAA04DgA2Vta93Y0tuMXcKBAACmDcEHytqhnetk2n34rOz+L6yAAYDZjeADZe+H29eoJhLWD7evYQUMAMwBNBlD2WtqiOW0Xj/c0aObKheo9rF2ba2L6tDOdSUcHQBgqsh8YFYxDcje7B9Q2paOne1nCgYAZhmCD5Stieo73IWoTMEAwOxC8IGyNVF9x6Gd67R/R7YWZH1ssda2vqS1rS+RBQGAWYCaD5St5o21zgZzbZ29OtzRo/WxxTrxzkcaSWUUCvh038olOna233nM4Y6enPoQAED5sWzbLqvuCYODg6qqqtLAwIAWLVpU6uGgTNxz4BX1xROyJE30Cxv0WWppXE0AAgAem8r5m8wHyk5bZ6+ean9bw8m0opEKXRgY1sKK8b+q4aBPiWQm57Zkxib7AQBljpoPlJ3DHT1KJNOyJfXFh5W2pXgiJWks6xEJB1VdGRr3WJ8lNqIDgDJH8IGyM1HwEPRZqomEtXvTKq2PLR53vyWR9QCAMse0C8rOyfOXx91WV1OlS0NX1byx1gkuCq2C8VnWjI8PAHBjCD5QdtyrVySpsb5wF9PmjbV6qv1tJZJp57ZkxlZbZy/ZDwAoY0y7oKzVRMK6e0V1wWZjTQ0xXU2lxz2GpmMAUN4IPlB2aiIVksaKRws1GzPdT1dHq3Iea0kFa0EAAOWDaReUneaNn3Gai7nrO9zNxoZGUoonkpKy0zIvnu2XrexqmBPvfFS6wQMArongA2Vj15HTau/ql8+ylMzYOnj8nLOjrQlCTLOxSDiomkjYyYy4G4+NFJiKAQCUD6ZdUBbaOnt17Gy/0na2aLSY5o21qomEdd/KJZKyK2OGRlIKB8d+lYeTGfZ4AYAyRvCBsvDkC90535teHsWceOcj9cUTau/qVzyRVHVlSPt3rJHfyk69UHQKAOWLaReUhfxkx2t77i943MHj5xRPJBUO+hQJBzWSSisU8Dv1ISfPX1Z7Vz9FpwBQxsh8oCyYFS5StqHYtSSSGQ0kkkokM6oMBZyakBPvfKS0ne0Vsrb1+IyNFwBw/ch8oCy8tueBSR23e9MqtbzQrbQ9ts/L5aGr2nXktE71XtHQSMo51uwHAwAoL2Q+MKs0NcTUun2NIuGgTCP1RDKtF8/2qy+eGFesuuvIae8HCQCYEMEHZp2mhpjOtDykfTvWOAFIsfUx+a3aAQClR/CBWaupIaZ9oytcAACzBzUfmNVMoekTR7sL3t9YH/VyOACASSD4wKznbsEu2eqLDyvos9TSuJrdbQGgDFm2bRdvJ1kCg4ODqqqq0sDAgBYtWlTq4QAAgEmYyvmbmg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOApgg8AAOCpQKkHkM9ssjs4OFjikQAAgMky521zHp9I2QUfH3/8sSRp+fLlJR4JAACYqo8//lhVVVUTHmPZkwlRPJTJZNTf36+FCxfKsqxSD8dTg4ODWr58ud5//30tWrSo1MOZs3ifvcH77B3ea2/wPk/Mtm19/PHHikaj8vkmruoou8yHz+fTpz/96VIPo6QWLVrEL7YHeJ+9wfvsHd5rb/A+F3etjIdBwSkAAPAUwQcAAPAUwUcZCYVCamlpUSgUKvVQ5jTeZ2/wPnuH99obvM/Tp+wKTgEAwNxG5gMAAHiK4AMAAHiK4AMAAHiK4AMAAHiK4KNM3X777bIsK+fjwIEDpR7WnPDcc8/p9ttvV0VFhTZs2KCTJ0+Wekhzyg9+8INxv7uf+9znSj2sOeHEiRPatm2botGoLMvS0aNHc+63bVtPPvmkbr31VoXDYT344IN69913SzPYWexa7/M3v/nNcb/jmzdvLs1gZymCjzL2wx/+UB988IHz8cgjj5R6SLPez3/+cz366KNqaWnR//zP/6i+vl6bNm3Shx9+WOqhzSmrV6/O+d199dVXSz2kOWFoaEj19fV67rnnCt7/4x//WIcOHdLzzz+vN954Q5WVldq0aZOGh4c9Hunsdq33WZI2b96c8zt+5MgRD0c4+5Vde3WMWbhwoZYtW1bqYcwpP/nJT/Sd73xH3/rWtyRJzz//vNrb2/VP//RP2rNnT4lHN3cEAgF+d2fAli1btGXLloL32batZ555Rk888YS2b98uSfrZz36mpUuX6ujRo/r617/u5VBntYneZyMUCvE7fgPIfJSxAwcO6KabbtK6dev0D//wD0qlUqUe0qx29epVnTp1Sg8++KBzm8/n04MPPqjXX3+9hCObe959911Fo1Hdcccd+sY3vqHf/e53pR7SnHf+/HlduHAh5/e7qqpKGzZs4Pd7BnR0dOiWW27RqlWr1NzcrEuXLpV6SLMKmY8ytWvXLv3Jn/yJqqur9d///d967LHH9MEHH+gnP/lJqYc2a/3hD39QOp3W0qVLc25funSpfvOb35RoVHPPhg0b9NOf/lSrVq3SBx98oNbWVv3Zn/2Zuru7tXDhwlIPb866cOGCJBX8/Tb3YXps3rxZX/nKV7RixQr19PTo+9//vrZs2aLXX39dfr+/1MObFQg+PLRnzx79/d///YTHvP322/rc5z6nRx991Lmtrq5OCxYs0F//9V/r6aefprUvypo7XV1XV6cNGzYoFovpX/7lX/Ttb3+7hCMDpod7Cuvzn/+86urqVFtbq46ODj3wwAMlHNnsQfDhob/7u7/TN7/5zQmPueOOOwrevmHDBqVSKf32t7/VqlWrZmB0c9/NN98sv9+vixcv5tx+8eJF5m5nUCQS0cqVK/Xee++Veihzmvkdvnjxom699Vbn9osXL2rt2rUlGtX8cMcdd+jmm2/We++9R/AxSQQfHlqyZImWLFlyXY89c+aMfD6fbrnllmke1fyxYMECrV+/Xi+//LJ27NghScpkMnr55Zf1N3/zN6Ud3Bz2ySefqKenR3/xF39R6qHMaStWrNCyZcv08ssvO8HG4OCg3njjDTU3N5d2cHPc73//e126dCkn6MPECD7K0Ouvv6433nhDX/rSl7Rw4UK9/vrr+tu//Vs1NTVp8eLFpR7erPboo4/q4Ycf1he+8AXdfffdeuaZZzQ0NOSsfsGN2717t7Zt26ZYLKb+/n61tLTI7/dr586dpR7arPfJJ5/kZJDOnz+vM2fOqLq6Wrfddpu++93vav/+/frsZz+rFStWaO/evYpGo06wjcmZ6H2urq5Wa2urvvrVr2rZsmXq6enR9773PX3mM5/Rpk2bSjjqWcZG2Tl16pS9YcMGu6qqyq6oqLDvvPNO+0c/+pE9PDxc6qHNCc8++6x922232QsWLLDvvvtuu7Ozs9RDmlO+9rWv2bfeequ9YMECu6amxv7a175mv/fee6Ue1pzwX//1X7akcR8PP/ywbdu2nclk7L1799pLly61Q6GQ/cADD9jnzp0r7aBnoYne5z/+8Y/2Qw89ZC9ZssQOBoN2LBazv/Od79gXLlwo9bBnFcu2bbtUgQ8AAJh/6PMBAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA8RfABAAA89f8DLciLnAPCWIkAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "plt.scatter(\n",
    "    embedding[:, 0],\n",
    "    embedding[:, 1],\n",
    "    s=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bafb8266-7451-495b-bdae-9bd51c6a214a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39babb0e-0fed-4bf0-95d2-0a6df7f614a5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
